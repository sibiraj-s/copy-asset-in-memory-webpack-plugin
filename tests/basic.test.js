const getCompiler = require('./helpers/getCompiler');
const compile = require('./helpers/compile');

const CompressionPlugin = require('compression-webpack-plugin');
const CopyAssetInMemoryPlugin = require('../src/index.js');

const getAssetNames = (stats) => {
  const { assets } = stats.compilation;
  return Object.keys(assets);
};

const getAssetSize = (stats, assetName) => {
  const { assets } = stats.compilation;
  return assets[assetName].size();
};

const hasAsset = (stats, assetName) => {
  const assets = getAssetNames(stats)
  if (assetName instanceof RegExp) {
    return assets.some(name => assetName.test(name))
  }

  return assets.includes(assetName)
}

it('should do nothing when there is no change in filename', async () => {
  const compiler = getCompiler();

  new CopyAssetInMemoryPlugin({
    test: /.js$/,
    to: (fileName) => fileName
  }).apply(compiler);

  const stats = await compile(compiler);

  const assets = getAssetNames(stats);

  expect(assets).toMatchSnapshot('assets');
  expect(assets.length).toBe(2);
});

it('should do copy assets to the new location for a given filter', async () => {
  const compiler = getCompiler();

  const to = (fileName) => `js/${fileName}`;

  new CopyAssetInMemoryPlugin({
    test: /.js$/,
    to,
  }).apply(compiler);

  const stats = await compile(compiler);

  const assets = getAssetNames(stats);

  expect(assets).toMatchSnapshot('assets');
  expect(hasAsset(stats, 'js/main.js')).toBeTruthy();
  expect(assets.length).toBe(3);
});

it('should do copy assets to the new location when destination is a string', async () => {
  const compiler = getCompiler();

  new CopyAssetInMemoryPlugin({
    test: /.js$/,
    to: 'js',
  }).apply(compiler);

  const stats = await compile(compiler);
  const assets = getAssetNames(stats);

  expect(assets).toMatchSnapshot('assets');
  expect(hasAsset(stats, 'js/main.js')).toBeTruthy();
  expect(assets.length).toBe(3);
});

it('should be able to transform file contents', async () => {
  const compiler = getCompiler();

  const newContent = 'new';
  const transform = (content) => `${content}${newContent}`;

  new CopyAssetInMemoryPlugin({
    test: /.js$/,
    to: (fileName) => `js/transformed-${fileName}`,
    transform,
  }).apply(compiler);

  const stats = await compile(compiler);

  const originalFileSize = getAssetSize(stats, 'main.js');
  const transformedFileSize = getAssetSize(stats, 'js/transformed-main.js');

  const assets = getAssetNames(stats);

  expect(assets).toMatchSnapshot('assets');
  expect(originalFileSize + newContent.length).toEqual(transformedFileSize);
  expect(assets.length).toBe(3);
});

it('should do copy assets to the new location and interpolate-name', async () => {
  const compiler = getCompiler();

  const newContent = 'new';
  const to = (fileName) => `js/[path]-${fileName}`;
  const transform = (content) => `${content}${newContent}`;

  new CopyAssetInMemoryPlugin({
    test: /.js$/,
    to,
    transform,
  }).apply(compiler);

  const stats = await compile(compiler);
  const assets = getAssetNames(stats);

  expect(assets).toMatchSnapshot('assets');
  expect(hasAsset(stats, 'js/[path]-main.js')).toBeFalsy();
  expect(assets.length).toBe(3);
});

it('should deleteOriginalAsset', async () => {
  const compiler = getCompiler();

  const to = (fileName) => `js/deleteOriginalAsset-${fileName}`;

  new CopyAssetInMemoryPlugin({
    test: /.*/,
    to,
    deleteOriginalAssets: true,
  }).apply(compiler);

  const stats = await compile(compiler);
  const assets = getAssetNames(stats);

  expect(assets).toMatchSnapshot('assets');
  expect(assets.length).toBe(2);
});

it('should add contenthash to the copied asset', async () => {
  const compiler = getCompiler();

  const to = () => `js/[name]-[contenthash:8].[ext]`;

  new CopyAssetInMemoryPlugin({
    test: /.js$/,
    to,
  }).apply(compiler);

  const stats = await compile(compiler);
  const assets = getAssetNames(stats);

  expect(assets).toMatchSnapshot('assets');
  expect(hasAsset(stats, /js\/file-[a-z-0-9]{0,8}.js$/)).toBeTruthy()
  expect(assets.length).toBe(3);
});

it('should work along with compression-webpack-plugin', async () => {
  const compiler = getCompiler();

  const to = (fileName) => `js/${fileName}`;

  new CopyAssetInMemoryPlugin({
    test: /.*.js(.map)?$/,
    to,
  }).apply(compiler);

  new CompressionPlugin({
    filename: '[file]',
    exclude: /js/,
    deleteOriginalAssets: 'keep-source-map',
    minRatio: 10,
  }).apply(compiler)

  const stats = await compile(compiler);
  const assets = getAssetNames(stats);

  expect(assets).toMatchSnapshot('assets');
  expect(hasAsset(stats, 'js/main.js')).toBeTruthy()
  expect(assets.length).toBe(4);
});
