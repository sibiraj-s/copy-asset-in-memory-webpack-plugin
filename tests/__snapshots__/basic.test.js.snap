// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add contenthash to the copied asset with given salt: assets 1`] = `
Array [
  "main.js",
  "main.js.map",
  "js/main-6f8c66dd.js",
]
`;

exports[`should add contenthash to the copied asset: assets 1`] = `
Array [
  "main.js",
  "main.js.map",
  "js/main-6f8c66dd.js",
]
`;

exports[`should be able to transform file contents: assets 1`] = `
Array [
  "main.js",
  "main.js.map",
  "js/transformed-main.js",
]
`;

exports[`should deleteOriginalAsset: assets 1`] = `
Array [
  "js/deleteOriginalAsset-main.js",
  "js/deleteOriginalAsset-main.js.map",
]
`;

exports[`should do copy assets to the new location and interpolate-name: assets 1`] = `
Array [
  "main.js",
  "main.js.map",
  "js/main.js-main.js",
]
`;

exports[`should do copy assets to the new location for a given filter: assets 1`] = `
Array [
  "main.js",
  "main.js.map",
  "js/main.js",
]
`;

exports[`should do copy assets to the new location when destination is a string: assets 1`] = `
Array [
  "main.js",
  "main.js.map",
  "js/main.js",
]
`;

exports[`should do nothing when there is no change in filename: assets 1`] = `
Array [
  "main.js",
  "main.js.map",
]
`;

exports[`should work along with compression-webpack-plugin: assets 1`] = `
Array [
  "main.js",
  "main.js.map",
  "js/main.js",
  "js/main.js.map",
]
`;
